File: inventory-service/tracing.js
const { NodeSDK } = require("@opentelemetry/sdk-node");
const { getNodeAutoInstrumentations } = require("@opentelemetry/auto-instrumentations-node");
const { OTLPTraceExporter } = require("@opentelemetry/exporter-trace-otlp-grpc");
const { resourceFromAttributes } = require("@opentelemetry/resources");

const exporter = new OTLPTraceExporter({
  url: process.env.OTLP_EXPORTER_ENDPOINT,
});

const sdk = new NodeSDK({
  traceExporter: exporter,
  resource: resourceFromAttributes({
    "service.name": process.env.SERVICE_NAME,
  }),
  instrumentations: [getNodeAutoInstrumentations()],
});

sdk.start();

File: notification-service/tracing.py
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry import trace
import os

resource = Resource.create({"service.name": os.getenv("SERVICE_NAME")})
provider = TracerProvider(resource=resource)
trace.set_tracer_provider(provider)

processor = BatchSpanProcessor(
    OTLPSpanExporter(endpoint=os.getenv("OTLP_EXPORTER_ENDPOINT"), insecure=True)
)
provider.add_span_processor(processor)
